README
NAME
    Test::Base - A Data Driven Testing Framework

SYNOPSIS
    A new test module:

        # lib/MyProject/Test.pm
        package MyProject::Test;
        use Test::Base -Base;

        use MyProject;

        package MyProject::Test::Filter;
        use Test::Base::Filter -base;

        sub my_filter {
            return MyProject->do_something(shift);
        }

    A sample test:

        # t/sample.t
        use MyProject::Test;

        plan tests => 1 * blocks;

        run_is input => 'expected';

        sub local_filter {
            s/my/your/;
        }

        __END__

        === Test one (the name of the test)
        --- input my_filter local_filter
        my
        input
        lines
        --- expected
        expected
        output

        === Test two
        This is an optional description
        of this particular test.
        --- input my_filter
        other
        input
        lines
        --- expected
        other expected
        output

DESCRIPTION
    Testing is usually the ugly part of Perl module authoring. Perl gives
    you a standard way to run tests with Test::Harness, and basic testing
    primitives with Test::More. After that you are pretty much on your own
    to develop a testing framework and philosophy. Test::More encourages you
    to make your own framework by subclassing Test::Builder, but that is not
    trivial.

    Test::Base gives you a way to write your own test framework base class
    that *is* trivial. In fact it is as simple as two lines:

        package MyTestFramework;
        use Test::Base -Base;

    A module called "MyTestFramework.pm" containing those two lines, will
    give all the power of Test::More and all the power of Test::Base to
    every test file that uses it. As you build up the capabilities of
    "MyTestFramework", your tests will have all of that power as well.

    "MyTestFramework" becomes a place for you to put all of your reusable
    testing bits. As you write tests, you will see patterns and duplication,
    and you can "upstream" them into "MyTestFramework". Of course, you don't
    have to subclass Test::Base at all. You can use it directly in many
    applications, including everywhere you would use Test::More.

    Test::Base concentrates on offering reusable data driven patterns, so
    that you can write tests with a minimum of code. At the heart of all
    testing you have inputs, processes and expected outputs. Test::Base
    provides some clean ways for you to express your input and expected
    output data, so you can spend your

          time focusing on that rather than your code scaffolding.

EXPORTED FUNCTIONS
    Test::Base extends Test::More and exports all of its functions. So you
    can basically write your tests the same as Test::More. Test::Base also
    exports many functions of its own:

    "is(actual, expected, [test-name])"
        This is the equivalent of Test::More's "is" function with one
        interesting twist. If your actual and expected results differ and
        the output is multi- line, this function will show you a unified
        diff format of output. Consider the benefit when looking for the one
        character that is different in hundreds of lines of output!

        Diff output requires the optional "Text::Diff" CPAN module. If you
        don't have this module, the "is()" function will simply give you
        normal Test::More output. To disable diffing altogether, set the
        "TEST_SHOW_NO_DIFFS" environment variable (or
        $ENV{TEST_SHOW_NO_DIFFS}) to a true value. You can also call the
        "no_diff" function as a shortcut.

    "blocks( [data-section-name] )"
        The most important function is "blocks". In list context it returns
        a list of "Test::Base::Block" objects that are generated from the
        test specification in the "DATA" section of your test file. In
        scalar context it returns the number of objects. This is useful to
        calculate your Test::More plan.

        Each Test::Base::Block object has methods that correspond to the
        names of that object's data sections. There is also a "name" and a
        "description" method for accessing those parts of the block if they
        were specified.

        The "blocks" function can take an optional single argument, that
        indicates to only return the blocks that contain a particular named
        data section. Otherwise "blocks" returns all blocks.

            my @all_of_my_blocks = blocks;

            my @just_the_foo_blocks = blocks('foo');

    "next_block()"
        You can use the next_block function to iterate over all the blocks.

            while (my $block = next_block) {
                ...
            }

        It returns undef after all blocks have been iterated over. It can
        then be called again to reiterate.

    "first_block()"
        Returns the first block or undef if there are none. It resets the
        iterator to the "next_block" function.

    "run(&subroutine)"
        There are many ways to write your tests. You can reference each
        block individually or you can loop over all the blocks and perform a
        common operation. The "run" function does the looping for you, so
        all you need to do is pass it a code block to execute for each
        block.

        The "run" function takes a subroutine as an argument, and calls the
        sub one time for each block in the specification. It passes the
        current block object to the subroutine.

            run {
                my $block = shift;
                is(process($block->foo), $block->bar, $block->name);
            };

    "run_is([data_name1, data_name2])"
        Many times you simply want to see if two data sections are
        equivalent in every block, probably after having been run through
        one or more filters. With the "run_is" function, you can just pass
        the names of any two data sections that exist in every block, and it
        will loop over every block comparing the two sections.

            run_is 'foo', 'bar';

        If no data sections are given "run_is" will try to detect them
        automatically.

        NOTE: Test::Base will silently ignore any blocks that don't contain
        both sections.

    "is_deep($data1, $data2, $test_name)"
        Like Test::More's "is_deeply" but uses the more correct Test::Deep
        module.

    "run_is_deeply([data_name1, data_name2])"
        Like "run_is_deeply" but uses "is_deep" which uses the more correct
        Test::Deep.

    "run_is_deeply([data_name1, data_name2])"
        Like "run_is" but uses "is_deeply" for complex data structure
        comparison.

    "run_is_deeply([data_name1, data_name2])"
        Like "run_is_deeply" but uses "is_deep" which uses the more correct
        Test::Deep.

    "run_like([data_name, regexp | data_name]);"
        The "run_like" function is similar to "run_is" except the second
        argument is a regular expression. The regexp can either be a "qr{}"
        object or a data section that has been filtered into a regular
        expression.

            run_like 'foo', qr{<html.*};
            run_like 'foo', 'match';

    "run_unlike([data_name, regexp | data_name]);"
        The "run_unlike" function is similar to "run_like", except the
        opposite.

            run_unlike 'foo', qr{<html.*};
            run_unlike 'foo', 'no_match';

    "run_compare(data_name1, data_name2)"
        The "run_compare" function is like the "run_is", "run_is_deeply" and
        the "run_like" functions all rolled into one. It loops over each
        relevant block and determines what type of comparison to do.

        NOTE: If you do not specify either a plan, or run any tests, the
        "run_compare" function will automatically be run.

    "delimiters($block_delimiter, $data_delimiter)"
        Override the default delimiters of "===" and "---".

    "spec_file($file_name)"
        By default, Test::Base reads its input from the DATA section. This
        function tells it to get the spec from a file instead.

    "spec_string($test_data)"
        By default, Test::Base reads its input from the DATA section. This
        function tells it to get the spec from a string that has been
        prepared somehow.

    "filters( @filters_list or $filters_hashref )"
        Specify a list of additional filters to be applied to all blocks.
        See "FILTERS" below.

        You can also specify a hash ref that maps data section names to an
        array ref of filters for that data type.

            filters {
                xxx => [qw(chomp lines)],
                yyy => ['yaml'],
                zzz => 'eval',
            };

        If a filters list has only one element, the array ref is optional.

    "filters_delay( [1 | 0] );"
        By default Test::Base::Block objects are have all their filters run
        ahead of time. There are testing situations in which it is
        advantageous to delay the filtering. Calling this function with no
        arguments or a true value, causes the filtering to be delayed.

            use Test::Base;
            filters_delay;
            plan tests => 1 * blocks;
            for my $block (blocks) {
                ...
                $block->run_filters;
                ok($block->is_filtered);
                ...
            }

        In the code above, the filters are called manually, using the
        "run_filters" method of Test::Base::Block. In functions like
        "run_is", where the tests are run automatically, filtering is
        delayed until right before the test.

    "filter_arguments()"
        Return the arguments after the equals sign on a filter.

            sub my_filter {
                my $args = filter_arguments;
                # is($args, 'whazzup');
                ...
            }

            __DATA__
            === A test
            --- data my_filter=whazzup

    "tie_output()"
        You can capture STDOUT and STDERR for operations with this function:

            my $out = '';
            tie_output(*STDOUT, $out);
            print "Hey!\n";
            print "Che!\n";
            untie *STDOUT;
            is($out, "Hey!\nChe!\n");

    "no_diff()"
        Turn off diff support for is() in a test file.

    "default_object()"
        Returns the default Test::Base object. This is useful if you feel
        the need to do an OO operation in otherwise functional test code.
        See OO below.

    "WWW() XXX() YYY() ZZZ()"
        These debugging functions are exported from the Spiffy.pm module.
        See Spiffy for more info.

    "croak() carp() cluck() confess()"
        You can use the functions from the Carp module without needing to
        import them. Test::Base does it for you by default.

TEST SPECIFICATION
    Test::Base allows you to specify your test data in an external file, the
    DATA section of your program or from a scalar variable containing all
    the text input.

    A *test specification* is a series of text lines. Each test (or block)
    is separated by a line containing the block delimiter and an optional
    test "name". Each block is further subdivided into named sections with a
    line containing the data delimiter and the data section name. A
    "description" of the test can go on lines after the block delimiter but
    before the first data section.

    Here is the basic layout of a specification:

        === <block name 1>
        <optional block description lines>
        --- <data section name 1> <filter-1> <filter-2> <filter-n>
        <test data lines>
        --- <data section name 2> <filter-1> <filter-2> <filter-n>
        <test data lines>
        --- <data section name n> <filter-1> <filter-2> <filter-n>
        <test data lines>

        === <block name 2>
        <optional block description lines>
        --- <data section name 1> <filter-1> <filter-2> <filter-n>
        <test data lines>
        --- <data section name 2> <filter-1> <filter-2> <filter-n>
        <test data lines>
        --- <data section name n> <filter-1> <filter-2> <filter-n>
        <test data lines>

    Here is a code example:

        use Test::Base;

        delimiters qw(### :::);

        # test code here

        __END__

        ### Test One
        We want to see if foo and bar
        are really the same...
        ::: foo
        a foo line
        another foo line

        ::: bar
        a bar line
        another bar line

        ### Test Two

        ::: foo
        some foo line
        some other foo line

        ::: bar
        some bar line
        some other bar line

        ::: baz
        some baz line
        some other baz line

    This example specifies two blocks. They both have foo and bar data
    sections. The second block has a baz component. The block delimiter is
    "###" and the data delimiter is ":::".

    The default block delimiter is "===" and the default data delimiter is
    "--- ".

    There are some special data section names used for control purposes:

        --- SKIP
        --- ONLY
        --- LAST

    A block with a SKIP section causes that test to be ignored. This is
    useful to disable a test temporarily.

    A block with an ONLY section causes only that block to be used. This is
    useful when you are concentrating on getting a single test to pass. If
    there is more than one block with ONLY, the first one will be chosen.

    Because ONLY is very useful for debugging and sometimes you forgot to
    remove the ONLY flag before committing to the VCS or uploading to CPAN,
    Test::Base by default gives you a diag message saying *I found ONLY ...
    maybe you're debugging?*. If you don't like it, use "no_diag_on_only".

    A block with a LAST section makes that block the last one in the
    specification. All following blocks will be ignored.

FILTERS
    The real power in writing tests with Test::Base comes from its filtering
    capabilities. Test::Base comes with an ever growing set of useful
    generic filters than you can sequence and apply to various test blocks.
    That means you can specify the block serialization in the most readable
    format you can find, and let the filters translate it into what you
    really need for a test. It is easy to write your own filters as well.

    Test::Base allows you to specify a list of filters to each data section
    of each block. The default filters are "norm" and "trim". These filters
    will be applied (in order) to the data after it has been parsed from the
    specification and before it is set into its Test::Base::Block object.

    You can add to the default filter list with the "filters" function. You
    can specify additional filters to a specific block by listing them after
    the section name on a data section delimiter line.

    Example:

        use Test::Base;

        filters qw(foo bar);
        filters { perl => 'strict' };

        sub upper { uc(shift) }

        __END__

        === Test one
        --- foo trim chomp upper
        ...

        --- bar -norm
        ...

        --- perl eval dumper
        my @foo = map {
            - $_;
        } 1..10;
        \ @foo;

    Putting a "-" before a filter on a delimiter line, disables that filter.

  Scalar vs List
    Each filter can take either a scalar or a list as input, and will return
    either a scalar or a list. Since filters are chained together, it is
    important to learn which filters expect which kind of input and return
    which kind of output.

    For example, consider the following filter list:

        norm trim lines chomp array dumper eval

    The data always starts out as a single scalar string. "norm" takes a
    scalar and returns a scalar. "trim" takes a list and returns a list, but
    a scalar is a valid list. "lines" takes a scalar and returns a list.
    "chomp" takes a list and returns a list. "array" takes a list and
    returns a scalar (an anonymous array reference containing the list
    elements). "dumper" takes a list and returns a scalar. "eval" takes a
    scalar and creates a list.

    A list of exactly one element works fine as input to a filter requiring
    a scalar, but any other list will cause an exception. A scalar in list
    context is considered a list of one element.

    Data accessor methods for blocks will return a list of values when used
    in list context, and the first element of the list in scalar context.
    This is usually "the right thing", but be aware.

  The Stock Filters
    Test::Base comes with large set of stock filters. They are in the
    "Test::Base::Filter" module. See Test::Base::Filter for a listing and
    description of these filters.

  Rolling Your Own Filters
    Creating filter extensions is very simple. You can either write a
    *function* in the "main" namespace, or a *method* in the
    "Test::Base::Filter" namespace or a subclass of it. In either case the
    text and any extra arguments are passed in and you return whatever you
    want the new value to be.

    Here is a self explanatory example:

        use Test::Base;

        filters 'foo', 'bar=xyz';

        sub foo {
            transform(shift);
        }

        sub Test::Base::Filter::bar {
            my $self = shift;       # The Test::Base::Filter object
            my $data = shift;
            my $args = $self->current_arguments;
            my $current_block_object = $self->block;
            # transform $data in a barish manner
            return $data;
        }

    If you use the method interface for a filter, you can access the block
    internals by calling the "block" method on the filter object.

    Normally you'll probably just use the functional interface, although all
    the builtin filters are methods.

    Note that filters defined in the "main" namespace can look like:

        sub filter9 {
            s/foo/bar/;
        }

    since Test::Base automatically munges the input string into $_ variable
    and checks the return value of the function to see if it looks like a
    number. If you must define a filter that returns just a single number,
    do it in a different namespace as a method. These filters don't allow
    the simplistic $_ munging.

OO
    Test::Base has a nice functional interface for simple usage. Under the
    hood everything is object oriented. A default Test::Base object is
    created and all the functions are really just method calls on it.

    This means if you need to get fancy, you can use all the object oriented
    stuff too. Just create new Test::Base objects and use the functions as
    methods.

        use Test::Base;
        my $blocks1 = Test::Base->new;
        my $blocks2 = Test::Base->new;

        $blocks1->delimiters(qw(!!! @@@))->spec_file('test1.txt');
        $blocks2->delimiters(qw(### $$$))->spec_string($test_data);

        plan tests => $blocks1->blocks + $blocks2->blocks;

        # ... etc

THE "TEST::BASE::BLOCK" CLASS
    In Test::Base, blocks are exposed as Test::Base::Block objects. This
    section lists the methods that can be called on a Test::Base::Block
    object. Of course, each data section name is also available as a method.

    "name()"
        This is the optional short description of a block, that is specified
        on the block separator line.

    "description()"
        This is an optional long description of the block. It is the text
        taken from between the block separator and the first data section.

    "seq_num()"
        Returns a sequence number for this block. Sequence numbers begin
        with 1.

    "blocks_object()"
        Returns the Test::Base object that owns this block.

    "run_filters()"
        Run the filters on the data sections of the blocks. You don't need
        to use this method unless you also used the "filters_delay"
        function.

    "is_filtered()"
        Returns true if filters have already been run for this block.

    "original_values()"
        Returns a hash of the original, unfiltered values of each data
        section.

SUBCLASSING
    One of the nicest things about Test::Base is that it is easy to
    subclass. This is very important, because in your personal project, you
    will likely want to extend Test::Base with your own filters and other
    reusable pieces of your test framework.

    Here is an example of a subclass:

        package MyTestStuff;
        use Test::Base -Base;

        our @EXPORT = qw(some_func);

        sub some_func {
            (my ($self), @_) = find_my_self(@_);
            ...
        }

        package MyTestStuff::Block;
        use base 'Test::Base::Block';

        sub desc {
            $self->description(@_);
        }

        package MyTestStuff::Filter;
        use base 'Test::Base::Filter';

        sub upper {
            $self->assert_scalar(@_);
            uc(shift);
        }

    Note that you don't have to re-Export all the functions from Test::Base.
    That happens automatically, due to the powers of Spiffy.

    The first line in "some_func" allows it to be called as either a
    function or a method in the test code.

DISTRIBUTION SUPPORT
    You might be thinking that you do not want to use Test::Base in you
    modules, because it adds an installation dependency. Fear not.
    Module::Install takes care of that.

    Just write a Makefile.PL that looks something like this:

        use inc::Module::Install;

        name            'Foo';
        all_from        'lib/Foo.pm';

        use_test_base;

        WriteAll;

    The line with "use_test_base" will automatically bundle all the code the
    user needs to run Test::Base based tests.

OTHER COOL FEATURES
    Test::Base automatically adds:

        use strict;
        use warnings;

    to all of your test scripts and Test::Base subclasses. A Spiffy feature
    indeed.

HISTORY
    This module started its life with the horrible and ridicule inducing
    name "Test::Chunks". It was renamed to "Test::Base" with the hope that
    it would be seen for the very useful module that it has become. If you
    are switching from "Test::Chunks" to "Test::Base", simply substitute the
    concept and usage of "chunks" to "blocks".

AUTHOR
    Ingy döt Net <ingy@cpan.org>

COPYRIGHT
    Copyright 2005-2014. Ingy döt Net.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See <http://www.perl.com/perl/misc/Artistic.html>



Changes
0.87 Tue Aug 19 16:35:34 PDT 2014
 - Replace tabs with spaces

0.86 Mon Aug 18 08:29:20 PDT 2014
 - Fix for issue/15

0.85 Mon Aug 18 08:07:52 PDT 2014
 - Pull Module::Install::TestBase out of Test::Base
 - It will be released in separate dist
 - Was causing Test::Base to fail tests

0.84 Sat Aug 16 15:08:32 PDT 2014
 - Remove .gitignore
 - Fix Meta error. s/zild/=zild/

0.83 Sat Aug 16 12:31:00 PDT 2014
 - Meta 0.0.2

0.82 Sat Aug 16 04:14:25 PDT 2014
 - Eliminate spurious trailing whitespace

0.81 Sat Aug 16 02:52:12 PDT 2014
 - Eliminate File::Basename from test/

0.80 Fri Aug 15 20:49:55 PDT 2014
 - Add t/000-require-modules.t

0.79 Sat Aug  9 00:55:07 PDT 2014
 - Dep on EU::MM 6.52

0.78 Sat Aug  9 00:23:58 PDT 2014
 - Fix swim errors

0.77 Fri Aug  8 12:39:43 PDT 2014
 - Change Provider test to use string eval, keeping it out of the eyes of
   Perl::Prereqs.

0.76 Thu Aug  7 00:29:39 PDT 2014
 - Fix bad encoding in Pod

0.75 Wed Aug  6 22:04:50 PDT 2014
 - Use PR/14 which makes old and new Test::Builders work. exodist++

0.74 Wed Aug  6 13:36:29 PDT 2014
 - Add dependency on Filter::Util::Call just to be sure
 - Spiffy requires it and we already require Spiffy

0.73 Wed Aug  6 13:24:45 PDT 2014
 - Doc fix. Thanks @Perlover++

0.72 Wed Aug  6 10:31:52 PDT 2014
 - Dep on new Spiffy-0.40 to get rid of warnings on blead 5.21.x

0.71 Wed Aug  6 09:50:04 PDT 2014
 - Applied PR/4 from schwern++

0.70 Wed Aug  6 09:30:03 PDT 2014
 - Fixed doc typo

0.69 Wed Aug  6 09:09:34 PDT 2014
 - Add $VERSION back into Test::Base

0.68 Wed Aug  6 08:56:09 PDT 2014
 - PR/11 and fixes

0.67 Tue Aug  5 10:16:09 PDT 2014
 - Add badges to doc

0.66 Mon Aug  4 00:39:44 PDT 2014
 - Remove (c) from Copyright
 - Ginish doc to Swim

0.65 Sat Aug  2 23:48:34 PDT 2014
 - Fix copyright years.

0.64 Fri Aug  1 23:16:11 PDT 2014
 - Add Algorithm::Diff and Text::Diff to test.requires

0.63 Fri Aug  1 23:04:30 PDT 2014
 - Switch to Zilla-Dist

0.62 Mon Feb 10 14:45:04 PST 2014
 - Bad plan in a test

0.61 Sat Feb  8 11:08:00 PST 2014
 - Switch to dzil

0.60 Mon Apr  4 15:51:09 CST 2011
 - Applied patch by andk++
 - Make regexp test skip on 5.14 for now.

0.59 Thu Aug 20 14:56:36 PDT 2009
 - Fixed test dep bug reported by Alias++.

0.58 Thu Mar 26 17:26:13 PDT 2009
 - Another undef filtering change.

0.57 Thu Mar 26 16:42:03 PDT 2009
 - Allow value of undef to be filtered without warnings.

0.56 Sat Mar  7 12:13:32 PST 2009
 - Add Test::Deep support with is_deep and run_is_deep

0.55 Thu Dec  4 01:10:11 PST 2008
 - Module::Install::TestBase::use_test_base require 'Filter::Util::Call'
   now.

0.54 Wed Nov 29 15:21:02 PST 2006
 - Make dependency on Filter::Util::Call explicit in Makefile.PL Thanks to
   Adriano Ferreira

0.53 Wed Nov 29 15:21:02 PST 2006
 - Changes from miyagawa and crew

0.52 Mon Jun 19 10:44:53 PDT 2006
 - Add use_ok to exports

0.51 Fri Jun 16 13:05:22 PDT 2006
 - Remove build-requires dep of Spiffy for Module::Install::TestBase
 - Add in a patch from the good folk at Socialtext.

0.50 Mon Jan 30 10:52:52 PST 2006
 - No change. 0.49 got borked on the way to CPAN

0.49 Mon Jan 30 10:52:48 PST 2006
 - Added Module::Install::TestBase

0.48 Sun Jan 29 10:19:46 PST 2006
 - Fixed test failures on windows

0.47 Thu Jan 19 10:59:37 PST 2006
 - Depend on newer Spiffy 0.29

0.46 Sat Jan 14 05:46:31 PST 2006
 - Don't sign the distribution tarball
 - Don't require the diffing stuff

0.45 Mon Jan  9 20:58:04 PST 2006
 - Let multilevel inheritance work!
 - no_diff function turns off diffing.

0.44 Fri Jul 22 23:38:04 PDT 2005
 - Bug fix in is_diff from rking
 - Allow Test::Base to be required without trying to run tests
 - allow ONLY|LAST|SKIP with run_* implicit names.

0.43 Sun Jun 19 03:14:40 PDT 2005
 - change Test::Base::Filter::block to current_block.
 - change Test::Base::Filter::arguments to current_arguments.
 - add split and Split filters
 - add join and Join filters
 - add reverse and Reverse filters
 - add hash filter
 - allow (parens) around a data section name for readability.
 - allow regexps on split
 - allow for compact, one-line data sections
 - allow for repeated filters
 - detect sections names automatically
 - import XXX stuff into Filter class
 - add run_compare
 - automatically set no_plan sometimes
 - automatically run run_compare if no plan set at END
 - massive refactoring of all tests

0.42 Tue Jun 14 09:31:25 PDT 2005
 - Make any block method callable with a dummy AUTOLOAD

0.41 Sun Jun 12 15:49:15 PDT 2005
 - Add first_block() function
 - Split Test::Base::Filter into a separate module

0.40 Sat Jun 11 20:55:42 PDT 2005
 - Change name from Test::Chunks to more lofty Test::Base
 - Change concept of "chunks" to "blocks"

0.38 Wed Jun  8 00:33:00 PDT 2005
 - Allow simple substitutions on $_ in filters defined in `main::`
 - Add a filter_arguments() function
 - Fixed a undef warning in `is()`

0.37 Tue Jun  7 11:04:07 PDT 2005
 - Implement rking style diff_is
 - Add filters `exec_perl_stdout`

0.36 Sun Jun  5 11:49:54 PDT 2005
 - add tie_output support
 - suppress warning in accessor
 - support backslash escapes in filter arguments
 - New filters `unchomp chop append eval_stdout eval_stderr eval_all`
 - Add join string to join filter
 - Add a Test-Less index

0.35 Thu Jun  2 17:46:30 PDT 2005
 - Subtle filter bug fixed

0.34 Sat May 28 23:55:49 PDT 2005
 - Allow "late" call of `filters`.
 - Allow for appending filters that are predefined.

0.33 Sat May 28 23:55:41 PDT 2005
 - Support `next_chunk` iterator.

0.32 Tue May 24 08:03:57 PDT 2005
 - Add a method to access filter arguments
 - Curry `use` args to Test::More
 - Change base64 filter to base64_decode base64_encode
 - Apply filter just before dispatch in run()
 - Apply filters in order
 - Default to Test::Chunks inline classes for subclassing modules (for
   Filter and Chunks)

0.31 Mon May 23 20:48:28 PDT 2005
 - Guess names for chunk_class and filter_class. Easier subclassing.

0.30 Mon May 23 16:39:23 PDT 2005

inc/Module/Install.pm
#line 1 "/Users/ingy/src/ingy/Test-Base/inc/Module/Install.pm - /Users/ingy/local/lib/perl5/site_perl/5.8.6/Module/Install.pm"
package Module::Install;

use 5.004;
use strict 'vars';
use vars qw{$VERSION};
BEGIN {
    # Don't forget to update Module::Install::Admin too!
    $VERSION = '0.54';
}

# inc::Module::Install must be loaded first
unless ( $INC{join('/', inc => split(/::/, __PACKAGE__)).'.pm'} ) {
    die <<"END_DIE";
Please invoke ${\__PACKAGE__} with:

    use inc::${\__PACKAGE__};

not:

    use ${\__PACKAGE__};

END_DIE
}

use Cwd        ();
use FindBin;
use File::Find ();
use File::Path ();

*inc::Module::Install::VERSION = *VERSION;
@inc::Module::Install::ISA     = 'Module::Install';

sub autoload {
    my $self   = shift;
    my $caller = $self->_caller;
    my $cwd    = Cwd::cwd();
    my $sym    = "$caller\::AUTOLOAD";

    $sym->{$cwd} = sub {
        my $pwd = Cwd::cwd();
        if ( my $code = $sym->{$pwd} ) {
            # delegate back to parent dirs
            goto &$code unless $cwd eq $pwd;
        }
        $$sym =~ /([^:]+)$/ or die "Cannot autoload $caller - $sym";
        unshift @_, ($self, $1);
        goto &{$self->can('call')} unless uc($1) eq $1;
    };
}

sub import {
    my $class = shift;
    my $self  = $class->new(@_);

    unless ( -f $self->{file} ) {
        require "$self->{path}/$self->{dispatch}.pm";
        File::Path::mkpath("$self->{prefix}/$self->{author}");
        $self->{admin} = 
          "$self->{name}::$self->{dispatch}"->new(_top => $self);
        $self->{admin}->init;
        @_ = ($class, _self => $self);
        goto &{"$self->{name}::import"};
    }

    *{$self->_caller . "::AUTOLOAD"} = $self->autoload;
    $self->preload;

    # Unregister loader and worker packages so subdirs can use them again
    delete $INC{"$self->{file}"};
    delete $INC{"$self->{path}.pm"};
}

sub preload {
    my ($self) = @_;

        unless ( $self->{extentions} ) {
                $self->load_extensions(
                        "$self->{prefix}/$self->{path}", $self
                        );
        }

    my @exts = @{$self->{extensions}};
    unless ( @exts ) {
        my $admin = $self->{admin};
        @exts = $admin->load_all_extensions;
    }

    my %seen_method;
    foreach my $obj ( @exts ) {
        while (my ($method, $glob) = each %{ref($obj) . '::'}) {
            next unless defined *{$glob}{CODE};
            next if $method =~ /^_/;
            next if $method eq uc($method);
            $seen_method{$method}++;
        }
    }

    my $caller = $self->_caller;
    foreach my $name (sort keys %seen_method) {
        *{"${caller}::$name"} = sub {
            ${"${caller}::AUTOLOAD"} = "${caller}::$name";
            goto &{"${caller}::AUTOLOAD"};
        };
    }
}

sub new {
    my ($class, %args) = @_;

    # ignore the prefix on extension modules built from top level.
    my $base_path = Cwd::abs_path($FindBin::Bin);
    unless ( Cwd::abs_path(Cwd::cwd()) eq $base_path ) {
        delete $args{prefix};
    }

    return $args{_self} if $args{_self};

    $args{dispatch} ||= 'Admin';
    $args{prefix}   ||= 'inc';
    $args{author}   ||= '.author';
    $args{bundle}   ||= 'inc/BUNDLES';
    $args{base}     ||= $base_path;

    $class =~ s/^\Q$args{prefix}\E:://;
    $args{name}     ||= $class;
    $args{version}  ||= $class->VERSION;

    unless ($args{path}) {
        $args{path}  = $args{name};
        $args{path}  =~ s!::!/!g;
    }
    $args{file}     ||= "$args{base}/$args{prefix}/$args{path}.pm";

    bless(\%args, $class);
}

sub call {
    my $self   = shift;
    my $method = shift;
    my $obj    = $self->load($method) or return;

    unshift @_, $obj;
    goto &{$obj->can($method)};
}

sub load {
    my ($self, $method) = @_;

    $self->load_extensions(
        "$self->{prefix}/$self->{path}", $self
    ) unless $self->{extensions};

    foreach my $obj (@{$self->{extensions}}) {
        return $obj if $obj->can($method);
    }

    my $admin = $self->{admin} or die <<"END_DIE";
The '$method' method does not exist in the '$self->{prefix}' path!
Please remove the '$self->{prefix}' directory and run $0 again to load it.
END_DIE

    my $obj = $admin->load($method, 1);
    push @{$self->{extensions}}, $obj;

    $obj;
}

sub load_extensions {
    my ($self, $path, $top_obj) = @_;

    unshift @INC, $self->{prefix}
        unless grep { $_ eq $self->{prefix} } @INC;

    local @INC = ($path, @INC);
    foreach my $rv ($self->find_extensions($path)) {
        my ($file, $pkg) = @{$rv};
        next if $self->{pathnames}{$pkg};

        local $@;
        my $new = eval { require $file; $pkg->can('new') };
        unless ( $new ) {
            warn $@ if $@;
            next;
        }
        $self->{pathnames}{$pkg} = delete $INC{$file};
        push @{$self->{extensions}}, &{$new}($pkg, _top => $top_obj );
    }

    $self->{extensions} ||= [];
}

sub find_extensions {
    my ($self, $path) = @_;

    my @found;
    File::Find::find( sub {
        my $file = $File::Find::name;
        return unless $file =~ m!^\Q$path\E/(.+)\.pm\Z!is;
        return if $1 eq $self->{dispatch};

        $file = "$self->{path}/$1.pm";
        my $pkg = "$self->{name}::$1"; $pkg =~ s!/!::!g;
        push @found, [ $file, $pkg ];
    }, $path ) if -d $path;

    @found;
}

sub _caller {
    my $depth  = 0;
    my $caller = caller($depth);

    while ($caller eq __PACKAGE__) {
        $depth++;
        $caller = caller($depth);
    }

    $caller;
}

1;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A Data Driven Testing Framework",
  "AUTHOR" => "Ingy d\x{f6}t Net <ingy\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Test-Base",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Test::Base",
  "PREREQ_PM" => {
    "Filter::Util::Call" => 0,
    "Spiffy" => "0.40"
  },
  "TEST_REQUIRES" => {
    "Algorithm::Diff" => "1.15",
    "ExtUtils::MakeMaker" => "6.52",
    "Text::Diff" => "0.35"
  },
  "VERSION" => "0.87",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Algorithm::Diff" => "1.15",
  "ExtUtils::MakeMaker" => "6.52",
  "Filter::Util::Call" => 0,
  "Spiffy" => "0.40",
  "Text::Diff" => "0.35"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



